features/0434-outofband/README.md

```diff
@@ -0,0 +1,417 @@
+
+- A user on a web browser wants to execute a protocol (for example, to issue a verifiable credential) between the website operator's agent and a user's agent. To enable the execution of the protocol, the two agents must have a connection. An out-of-band message is used to trigger the creation of a new, or the reuse an existing, connection.
+- A website operator wants to request something from a user on their site where there is no (known) connection with the user.
+
+In both cases, there is only a single out of band protocol message sent. The message responding to the out of band message is a DIDComm message from an appropriate protocol.
+
+Note that the website-to-agent model is not the only such interaction enabled by the out of band protocol, and a QR code is not the only delivery mechanism for out of band messages. They are useful as examples of the purpose of the protocol.
+
+### Roles
+
+The out of band protocol has two roles: __sender__ and __receiver__.
+
+#### sender
+
+The agent that generates the out of band message and makes it available to the other party.
+
+#### receiver
+
+The agent that receives the out of band message and decides how to respond. There is no out of band protocol message with which the receiver will respond. Rather, if they respond, they will use a message from another protocol that the sender will understand.
+
+### States
+
+The state machines for the sender and receiver are a bit odd for the out of band protocol because of it being a single message protocol with a response from a different protocol. In the following state machine diagrams we generically describe the response message from the *receiver* as being a DIDComm message without associating the message with a specific protocol.
+
+The *sender* state machine is as follows:
+
+[![sender state machine google doc](state-machine-sender.png)](https://docs.google.com/spreadsheets/d/1o1szGCMx6FzrxfAKV76IEYmeKRJqvV3gd1r3Qfv1Q9o/edit#gid=1176419697)
+
+Note the "optional" reference under the second event in the `await-response` state. That is to indicate that an out of band message might be a single use message with a transition to done, or reusable message (received by many receivers) with a transition back to `await-response`.
+
+The *receiver* state machine is as follows:
+
+[![receiver state machine google doc](state-machine-receiver.png)](https://docs.google.com/spreadsheets/d/1o1szGCMx6FzrxfAKV76IEYmeKRJqvV3gd1r3Qfv1Q9o/edit#gid=10874521)
+
+Worth noting is the first event of the `done` state, where the receiver may receive the message multiple times. This represents, for example, an agent returning to the same website and being greeted with the same QR code each time.
+
+### Messages
+
+There are two messages in the out of band protocol. They are both sent by the *sender*. Only one of the message types is ever used in an instance of the protocol.
+
+#### Message Type: `https://didcomm.org/outofband/VER/invitation`
```

Comment:
I feel like we are missing out on an opportunity to generalize a little bit further. We have here an invitation message, but we are assuming that the invitation is always an invitation to connect. Why couldn't it be an invitation to do X, Y, or Z instead, where connecting is a precondition? We could do this by having a new property telling what the goal of the invitation is:

```jsonc
{
    // all the stuff that's already in an invitation goes here, including the existing
    // protocols field. Then we add:

    "goal": [
        // Each item in the "goal" array is an alternative "weighted protocol start spec":
        //     protocol = the protocol to start
        //     your_role = named role the recipient should take
        //     preference = value between 0 and 1, expressing how preferred the option is.
        //         This allows two options to be weighted the same from the inviter's perspective.
        {"protocol": "https://didcomm.org/issue-credential/2.0", "your_role": "issuer", "preference": 0.8}
        {"protocol": "https://didcomm.org/issue-credential/1.3", "your_role": "issuer", "preference": 0.5}
        {"protocol": "https://didcomm.org/cred-iss/0.9", "your_role": "creator", "preference": 0.2}
    ]
}
```
The recipient decides whether to connect based on whether they agree to the goal, which is better than just connecting for no reason (gives the recipient better reason to filter spam). And the recipient can see which way the goal will be accomplished, so they'll know whether the connection is DOA because they can't accomplish the goal. Also, now you have a way to explicitly tell the recipient to start the protocol, or to take a passive role and let the inviter start the protocol as soon as the connection is established. You also have a way to start protocols with more than 2 roles (e.g., auctioneer, bidder, referee).

If we also made the `protocols` array into a `connect` array, and gave it the same structure, we could solve the prioritization problem, and we get the flexibility to do funkier connection types (e.g., let's connect by asking Alice to introduce us). Without this, you can set up a connection, but you're hardwired into the assumption that the invitee is the requester, and into the assumption that after the connection is set up, the inviter will make the first move in the subsequent protocol.

Reply:
We discussed this and decided while potentially useful, it was more complex than necessary.  It was in a first cut, but dropped during a discussion about the resulting benefits.

The purpose of the sender getting the "pthid" of the follow on message is so that they can know the context of the invitation that they initiated for use after the connection is established.  With that information, they can initiate the follow on action, whatever that might be. In many cases - particularly in a website/QR code scenario, both parties will know what will come next.

Further, by keeping that complexity out of the message, we keep the out of band message smaller.

Reply:
Ok - I somehow missed this giant point. My bad.

The main driver for this protocol on our end was to kickstart workflows *for long term relationships*. So we need did-exchange + something that describes whatever follows next. I kind of assumed the `request` message was setup for this but it appears to be tailored to ephemeral challenges?

Reply:
Apparently I'm struggling to make my point. Let me try again. I have three reasons why I'm recommending a change.

1. Spam

    Connecting, as we have done it thus far in the Connection and DID Exchange protocols, is often a bad goal, in and of itself. I get a steady stream of "let's connect" messages on LinkedIn, and I have to sort through them and decide whether I think a connection is a good idea. Usually, the way I decide is by looking at the specifics of the sender's intentions. A message saying, "Let's connect so we can collaborate on X, as we discussed in conversation Y" is a lot better than a messaging saying "let's connect." The latter is often spam. Spam with agents will be a thing.

    What we have in the current design is a protocol for proposing connections *with no further context*. This is problematic from a spam perspective. We need to be able to specify intentions beyond connection, or else we'll just have to invent a new out of band protocol later that has that feature. Spam will demand it.

2. Who starts the next interaction

    You said "The purpose of the sender getting the `pthid` of the follow on message is so that they can know the context of the invitation that they initiated for use after the connection is established. With that information, they can initiate the follow on action..." This is all well and good. But what if the follow-on action should be begun *by the recipient*? How does Alice say to Bob, "Hey, let's connect, so *you* can bid on my Rembrandt that I'm auctioning?" Right now we only support, "Hey, let's connect, so *I* can bid on your Rembrandt that you're auctioning."

3. Deciding whether the next interaction will be possible

    Building on the previous example, suppose the next planned interaction is auction bidding. Suppose that Alice only supports auction 2.0 protocol, and Bob only supports auction 1.0 protocol. It is likely that from the recipient's perspective, there is no point in connecting if the true goal for the connection (the auction, not just connecting) will be frustrated. Therefore, the OOB protocol ought to let the sender express the eventual goal, so that can be taken into account by the recipient when evaluating the desirability of connecting.

I'm not saying that every connection must have a further goal; just having a connection might be the goal. In that case, the `goal` array I proposed would be blank. What I'm saying is that we ought to make it possible to tell the difference between a connection without a goal, and a connection with a goal. This helps with spam; it helps make the next interaction one startable by other parties besides the sender; it helps a recipient decide if the connection is a dead end.

And I'm saying that some OOB messages might not have the goal of connecting at all; it's possible that the goal is to complete a connectionless interaction, in which case the `protocol` array (perhaps better renamed `connection_protocol` array) would be blank. Then the second message type you're imagining goes away. The OOB protocol collapses to one message type that supports one additional field.

Reply:
I get your argument, but I don't find it compelling.  As I said, we did talk about this and decided it was more complicated than worthwhile.

On the spam, the invitee has to take a step to get the out of band message.  I don't know how the inviter gets the unsolicited message.  Is that via email?  Either way, the invitee has already taken a step by accepting the OOB message.

On the initiator, I think in the majority of cases the inviter will initiate the next action. If not, I think rather than making OOB invitation more complex, we add a DIDComm message that supports that.  Andrew's "Action Menu" is our WIP on that - https://hackmd.io/jrWg2kw7QuWx9XS0R3_XPQ

The deciding if the next interaction is possible is IMHO another "too complex" issue for invitation.

That all said, George is now around to your side of the argument, so we'll have to figure out how to break through this.

We really need to nail this down soon as it is a very high priority item to get moving in the community. The lack of reuse of connections is painful.

Reply:
I should mention that further _features_ don't need a new protocol... a new version will do.

On spam: The primary fighting mechanism is indeed context, but that context is provided by a pthid of an invitation. A public DID can be used to jump into didexchange directly, but that method will be highly subject to spam.

Providing the 'goals' does add some color as to purpose, but I believe having a pthid of an outofband message will be the most useful method.

Reply:
On inviting the remote party (recipient of outofband message) to start the interaction: I think there are a few ways to do this.
1. Include a message you want the recipient to send back to you. This was suggested by Andrew, but I think very dangerous. The need to strictly validate is high.
2. Include a protocol and role. I could be convinced of this over 1. Same value, much less dangerous. I still am voting that this becomes a future version feature.
3. Use Pthid as context to prompt an interaction. Sending a BidRequest message isn't much different than requesting the recipient play the role of bidder. I think this method has the advantage of making agents easier to develop in the short term.

Reply:
I don't understand the reasoning on a pthid providing the link. That context-building capability is only useful in retrospect; that is, you can't consider the context when you're evaluating the invitation; you can only connect it later, after the connection is made. No spam benefits. That feels helpful to me, but WAY less helpful than setting the context at the outset. I'm really pretty strongly misaligned with this. I think we're making a mistake to treat the spam problem so casually.

Reply:
The claim I'm making is that the `pthid` provides the link to the context of the invitation in the inviter's context - why it was created. It tells the invitee nothing.  So the inviter does not need anything else.

Further, I'm claiming that the invitee knows the context of the claim before they process the invitation because the had to do something to ingest the invitation - scan it, click on it, etc.  At most, in the invitation they get at most self-attested information from the inviter - a label, and if you win, some sort of indication of the purpose of the invitation.  My view  is that I don't see the benefit of the extra (self-attested) information within the invitation as spam prevention.

Reply:
Right. I agree with your logic on the sender knowing the context via the pthid.

I disagree with your logic on the invitee knowing the context. I don't agree that we can assume the invitee had to take a human action to receive the OOB message. I gave several examples of how this could not be true, in my writeup about the spam concern.